app/jobs/notifications/remove_all_by_action_job.rb

```diff
@@ -0,0 +1,11 @@
+module Notifications
+  class RemoveAllByActionJob < ApplicationJob
+    queue_as :remove_all_by_action_notifications
+
+    def perform(notifiable_ids, notifiable_type, action, service = Notifications::RemoveAllByAction)
+      return unless %w[Article Comment Mention].include?(notifiable_type)
+
+      service.call(Array.wrap(notifiable_ids), notifiable_type, action)
```

Comment:
@rhymes @lightalloy I ended up adding the guard clause in the job, and used `return` instead of `raise` so the job wouldn't attempt to run multiple times. But maybe it should fail loudly?

Also, I ended up adding it in the job, but maybe the logic should be in the service itself?

Reply:
You can also use `before_enqueue` callback in the job class. You can do the check in the hook and raise to cancel the job from being enqueued.

Reply:
Maybe we can exit the job if `notifiable_ids` is empty
Unfortunately Rails will perform a SQL query if the array is empty:

```ruby
[1] pry(main)> Notification.where(id: [])
  Notification Load (69.6ms)  SELECT "notifications".* FROM "notifications" WHERE 1=0
```

Not sure if it should be here or inside the service call, what do you think @lightalloy?

Or we can just avoid putting the job in the queue in the first place if there are no related objects

Reply:
Actually, I wouldn't mind these checks in all the places.
First, I would put the check to avoid queueing the job, not to waste the resources on queueing/executing the job. But the checks inside the job (and event in the service) make sense to me too.